src/algorithms.hpp

```diff
@@ -283,43 +285,101 @@ bool is_unique(const C& c)
     return std::adjacent_find(c.begin(), c.end()) == c.end();
 }
 
+template <typename R1, typename R2>
+class index_into_iterator:
+    public std::iterator<std::forward_iterator_tag, typename R1::value_type>
+{
+public:
+    using value_type = typename R1::value_type;
+
+private:
+    using super_iterator = typename R1::const_iterator;
+    using sub_iterator = typename R2::const_iterator;
+
+    mutable super_iterator super_it_;
+    const super_iterator super_end_;
```

Comment:
I agree.
* `index_into_iterator` is now templated on the 3 iterator types
* updated unit test to check that arrays work specifically